From 029e53cdfaf064da818442bb5674f8d6ea4a2f6b Mon Sep 17 00:00:00 2001
From: Hariharan Ramanathan <quic_hramanat@quicinc.com>
Date: Thu, 24 Apr 2025 15:34:14 +0530
Subject: [PATCH 05/15] UPSTREAM:DP:ath12k:add statistic support for
counting wmm sent packets

Segregated wmm types for rx path and tx path introduced to
handle bidirectional traffic flow. This will avoid misplaced
piling of the wmm buckets.

Original Patch:
716-ath12k-sent-pkts.patch

Signed-off-by: saleemuddin shaik <quic_saleemud@quicinc.com>
Signed-off-by: Hariharan Ramanathan <quic_hramanat@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h    |  8 ---
 drivers/net/wireless/ath/ath12k/debugfs.c | 68 +++++++++++++++++++++++
 drivers/net/wireless/ath/ath12k/dp.h      | 14 +++++
 drivers/net/wireless/ath/ath12k/dp_htt.c  |  2 +
 drivers/net/wireless/ath/ath12k/dp_htt.h  |  2 +
 drivers/net/wireless/ath/ath12k/mac.c     | 15 +++++
 6 files changed, 101 insertions(+), 8 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 2583824..d277a1a 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -76,14 +76,6 @@ enum ath12k_bdf_search {
 	ATH12K_BDF_SEARCH_BUS_AND_BOARD,
 };
 
-enum wme_ac {
-	WME_AC_BE,
-	WME_AC_BK,
-	WME_AC_VI,
-	WME_AC_VO,
-	WME_NUM_AC
-};
-
 #define ATH12K_HT_MCS_MAX	7
 #define ATH12K_VHT_MCS_MAX	9
 #define ATH12K_HE_MCS_MAX	11
diff --git a/drivers/net/wireless/ath/ath12k/debugfs.c b/drivers/net/wireless/ath/ath12k/debugfs.c
index 7e74c42..46e2a8e 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -8,6 +8,7 @@
 #include "dp_tx.h"
 #include "debug.h"
 #include "debugfs.h"
+#include "debug.h"
 #include "debugfs_htt_stats.h"
 
 static ssize_t ath12k_write_simulate_radar(struct file *file,
@@ -1411,6 +1412,69 @@ void ath12k_debugfs_fw_stats_register(struct ath12k *ar)
 	ath12k_fw_stats_init(ar);
 }
 
+static ssize_t ath12k_read_wmm_stats(struct file *file,
+                                    char __user *ubuf,
+                                    size_t count, loff_t *ppos)
+{
+       struct ath12k *ar = file->private_data;
+       struct ath12k_base *ab = ar->ab;
+       struct ath12k_dp *dp = ath12k_ab_to_dp(ab);
+       struct ath12k_pdev_dp *dp_pdev = NULL;
+       int len = 0;
+       int size = 2048;
+       char *buf;
+       ssize_t retval;
+       u64 total_wmm_sent_pkts = 0;
+
+       if (!dp) {
+	       ath12k_warn(ar->ab, "ath12k_dp not present%s\n", __func__);
+	       return -ENOMEM;
+       }
+
+       dp_pdev = dp->dp_pdevs[ar->pdev_idx];
+
+       if (!dp_pdev) {
+	       ath12k_warn(dp, "dp_pdev not present for pdev_idx %d in %s\n",
+			   ar->pdev_idx, __func__);
+	       return -ENOMEM;
+       }
+
+       buf = kzalloc(size, GFP_KERNEL);
+       if (!buf) {
+               ath12k_warn(dp, "failed to allocate the buffer%s\n", __func__);
+               return -ENOMEM;
+       }
+
+       wiphy_lock(dp_pdev->hw->wiphy);
+       for (count = 0; count < WME_NUM_AC; count++)
+               total_wmm_sent_pkts += dp_pdev->wmm_stats.total_wmm_tx_pkts[count];
+
+       len += scnprintf(buf + len, size - len, "Total number of wmm_sent: %llu\n",
+                        total_wmm_sent_pkts);
+       len += scnprintf(buf + len, size - len, "Num of BE wmm_sent: %llu\n",
+                        dp_pdev->wmm_stats.total_wmm_tx_pkts[WME_AC_BE]);
+       len += scnprintf(buf + len, size - len, "Num of BK wmm_sent: %llu\n",
+                        dp_pdev->wmm_stats.total_wmm_tx_pkts[WME_AC_BK]);
+       len += scnprintf(buf + len, size - len, "Num of VI wmm_sent: %llu\n",
+                        dp_pdev->wmm_stats.total_wmm_tx_pkts[WME_AC_VI]);
+       len += scnprintf(buf + len, size - len, "Num of VO wmm_sent: %llu\n",
+                        dp_pdev->wmm_stats.total_wmm_tx_pkts[WME_AC_VO]);
+
+       wiphy_unlock(dp_pdev->hw->wiphy);
+
+       if (len > size)
+               len = size;
+       retval = simple_read_from_buffer(ubuf, count, ppos, buf, len);
+       kfree(buf);
+
+       return retval;
+}
+
+static const struct file_operations fops_wmm_stats = {
+       .read = ath12k_read_wmm_stats,
+       .open = simple_open,
+};
+
 void ath12k_debugfs_register(struct ath12k *ar)
 {
 	struct ath12k_base *ab = ar->ab;
@@ -1440,6 +1504,10 @@ void ath12k_debugfs_register(struct ath12k *ar)
 			    ar, &fops_tpc_stats_type);
 	init_completion(&ar->debug.tpc_complete);
 
+	debugfs_create_file("wmm_stats", 0644,
+		            ar->debug.debugfs_pdev, ar,
+			    &fops_wmm_stats);
+
 	ath12k_debugfs_htt_stats_register(ar);
 	ath12k_debugfs_fw_stats_register(ar);
 
diff --git a/drivers/net/wireless/ath/ath12k/dp.h b/drivers/net/wireless/ath/ath12k/dp.h
index a2193c2..0cc5f8d 100644
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -108,6 +108,19 @@ enum ath12k_dp_ppdu_state {
 	DP_PPDU_STATUS_DONE,
 };
 
+enum wme_ac {
+	WME_AC_BE,
+	WME_AC_BK,
+	WME_AC_VI,
+	WME_AC_VO,
+	WME_NUM_AC
+};
+
+struct ath12k_wmm_stats {
+       int tx_type;
+       u64 total_wmm_tx_pkts[WME_NUM_AC];
+};
+
 struct dp_mon_mpdu {
 	struct list_head list;
 	struct sk_buff *head;
@@ -159,6 +172,7 @@ struct ath12k_pdev_dp {
 
 	struct ieee80211_rx_status rx_status;
 	struct ath12k_mon_data mon_data;
+	struct ath12k_wmm_stats wmm_stats;
 };
 
 #define DP_NUM_CLIENTS_MAX 64
diff --git a/drivers/net/wireless/ath/ath12k/dp_htt.c b/drivers/net/wireless/ath/ath12k/dp_htt.c
index 31423c7..3d12699 100644
--- a/drivers/net/wireless/ath/ath12k/dp_htt.c
+++ b/drivers/net/wireless/ath/ath12k/dp_htt.c
@@ -299,6 +299,8 @@ ath12k_update_per_peer_tx_stats(struct ath12k_pdev_dp *dp_pdev,
 					  HTT_PPDU_STATS_ACK_BA_INFO_NUM_MSDU_M);
 		tid = le32_get_bits(usr_stats->ack_ba.info,
 				    HTT_PPDU_STATS_ACK_BA_INFO_TID_NUM);
+		dp_pdev->wmm_stats.tx_type = ath12k_tid_to_ac(tid > ATH12K_DSCP_PRIORITY ? 0: tid);
+		dp_pdev->wmm_stats.total_wmm_tx_pkts[dp_pdev->wmm_stats.tx_type]++;
 	}
 
 	if (common->fes_duration_us)
diff --git a/drivers/net/wireless/ath/ath12k/dp_htt.h b/drivers/net/wireless/ath/ath12k/dp_htt.h
index c4f66b3..56dae4a 100644
--- a/drivers/net/wireless/ath/ath12k/dp_htt.h
+++ b/drivers/net/wireless/ath/ath12k/dp_htt.h
@@ -36,6 +36,8 @@ struct ath12k_dp;
 
 #define HTT_TX_WBM_COMP_INFO2_ACK_RSSI		GENMASK(31, 24)
 
+#define ATH12K_DSCP_PRIORITY 7
+
 struct htt_tx_wbm_completion {
 	__le32 rsvd0[2];
 	__le32 info0;
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 8983578..69eb546 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1375,12 +1375,27 @@ int ath12k_mac_vdev_stop(struct ath12k_link_vif *arvif)
 {
 	struct ath12k_vif *ahvif = arvif->ahvif;
 	struct ath12k *ar = arvif->ar;
+	struct ath12k_dp *dp = ath12k_ab_to_dp(ar->ab);
+	struct ath12k_pdev_dp *dp_pdev = NULL;
 	int ret;
 
+	if (!dp) {
+		ath12k_err(ar->ab, "ath12k_dp not present%s",__func__);
+		goto err;
+	}
+
+	dp_pdev = dp->dp_pdevs[ar->pdev_idx];
+
+	if (!dp_pdev) {
+		ath12k_err(ar->ab, "dp_pdev not present%s",__func__);
+		goto err;
+	}
 	lockdep_assert_wiphy(ath12k_ar_to_hw(ar)->wiphy);
 
 	reinit_completion(&ar->vdev_setup_done);
 
+	memset(&dp_pdev->wmm_stats, 0, sizeof(struct ath12k_wmm_stats));
+
 	ret = ath12k_wmi_vdev_stop(ar, arvif->vdev_id);
 	if (ret) {
 		ath12k_warn(ar->ab, "failed to stop WMI vdev %i: %d\n",
-- 
2.34.1

